odl/discr/discr_ops.py

```diff
@@ -134,30 +139,71 @@ def finite_diff(f, out=None, axis=0, dx=1.0, edge_order=2,
+        # Numerical differentiation: 2nd order interior
+        slice_out[axis] = slice(1, -1)
+        slice_node1[axis] = slice(2, None)
+        slice_node2[axis] = slice(None, -2)
+        # 1D equivalent: out[1:-1] = (f[2:] - f[:-2])/2.0
+        np.subtract(f_data[slice_node1], f_data[slice_node2], out[slice_out])
+        out[slice_out] /= 2.0
+    elif method == 'forward':
+        # Numerical differentiation: 2nd order interior
+        slice_out[axis] = slice(1, -1)
+        slice_node1[axis] = slice(2, None)
+        slice_node2[axis] = slice(1, -1)
+        # 1D equivalent: out[1:-1] = (f[2:] - f[1:-1])
+        np.subtract(f_data[slice_node1], f_data[slice_node2], out[slice_out])
+    elif method == 'backward':
+        # Numerical differentiation: 2nd order interior
+        slice_out[axis] = slice(1, -1)
+        slice_node1[axis] = slice(1, -1)
+        slice_node2[axis] = slice(None, -2)
+        # 1D equivalent: out[1:-1] = (f[1:-1] - f[:-2])
+        np.subtract(f_data[slice_node1], f_data[slice_node2], out[slice_out])
 
     # central differences
     if zero_padding:
         # Assume zeros for indices outside the domain of `f`
 
-        slice_out[axis] = 0
-        slice_node1[axis] = 1
-        # 1D equivalent: out[0] = (f[1] - 0)/2.0
-        out[slice_out] = f_data[slice_node1] / 2.0
+        if method == 'central':
+            slice_out[axis] = 0
+            slice_node1[axis] = 1
+            # 1D equivalent: out[0] = (f[1] - 0)/2.0
+            out[slice_out] = f_data[slice_node1] / 2.0
 
-        slice_out[axis] = -1
-        slice_node2[axis] = -2
-        # 1D equivalent: out[-1] = (0 - f[-2])/2.0
-        out[slice_out] = - f_data[slice_node2] / 2.0
+            slice_out[axis] = -1
+            slice_node2[axis] = -2
+            # 1D equivalent: out[-1] = (0 - f[-2])/2.0
+            out[slice_out] = - f_data[slice_node2] / 2.0
+        elif method == 'forward':
+            slice_out[axis] = 0
+            slice_node1[axis] = 1
+            slice_node2[axis] = 0
+            # 1D equivalent: out[0] = (f[1] - f[0])
+            out[slice_out] = f_data[slice_node1] - f_data[slice_node2]
+
+            slice_out[axis] = -1
+            slice_node2[axis] = -1
+            # 1D equivalent: out[-1] = (0 - f[-1])
+            out[slice_out] = - f_data[slice_node2]
+        elif method == 'backward':
+            slice_out[axis] = 0
+            slice_node1[axis] = 0
+            # 1D equivalent: out[0] = (f[0] - 0)
+            out[slice_out] = f_data[slice_node1]
+
+            slice_out[axis] = -1
+            slice_node1[axis] = -1
+            slice_node2[axis] = -2
+            # 1D equivalent: out[-1] = (f[-1] - f[-2])
+            out[slice_out] = f_data[slice_node1] - f_data[slice_node2]
 
     # one-sided differences
     else:
+        if method != 'central':
+            raise NotImplementedError('Can only use zero padding with non-'
+                                      'central differences')
```

Comment:
We could also make the arguments a bit more structured i.e. instead of `zero_padding` using `boundary_method` with `zero_padding` or `one-sided` as options?